language/move-vm/types/src/values/values_impl.rs

```diff
@@ -178,24 +178,30 @@ pub struct Struct {
     fields: Vec<ValueImpl>,
 }
 
-/// A special value that lives in global storage.
-///
-/// Callers are allowed to take global references from a `GlobalValue`. A global value also contains
-/// an internal flag, indicating whether the value has potentially been modified or not.
-///
-/// For any given value in storage, only one `GlobalValue` may exist to represent it at any time.
-/// This means that:
-/// * `GlobalValue` **does not** and **should not** implement `Clone`!
-/// * a borrowed reference through `borrow_global` is represented through a `&GlobalValue`.
-/// * `borrow_global_mut` is also represented through a `&GlobalValue` -- the bytecode verifier
-///   enforces mutability restrictions.
-/// * `move_from` is represented through an owned `GlobalValue`.
+/// A special "slot" in global storage that can hold a resource. It also keeps track of the status
+/// of the resource relative to the global state, which is necessary to compute the effects to emit
+/// at the end of transaction execution.
 #[derive(Debug)]
-pub struct GlobalValue {
-    status: Rc<RefCell<GlobalDataStatus>>,
-    fields: Rc<RefCell<Vec<ValueImpl>>>,
+enum GlobalValueImpl {
+    /// No resource resides in this slot or in storage.
+    None,
+    /// A resource has been published to this slot and it did not previously exist in storage.
+    Fresh { fields: Rc<RefCell<Vec<ValueImpl>>> },
+    /// A resource resides in this slot and also in storage. The status flag indicates whether
+    /// it has potentially been altered.
+    Cached {
+        fields: Rc<RefCell<Vec<ValueImpl>>>,
+        status: Rc<RefCell<GlobalDataStatus>>,
```

Comment:
Is this ever mutated after a `GlobalValueImpl` is created? As far as I can tell, it is set to `Clean` upon creation and never changed.

Reply:
Yes, see my other comments.